### ISSUE TITLE
Implement language server protocol

### ISSUE THREAD

**Issue by lpil** [gleam-lang/gleam#180]
To enable IDE like features in all editors.

**Comment by gamebox** [comment]
I will take this on after #157 - which I'm happy to help with.

**Comment by gamebox** [comment]
You can assign this to me as well

**Comment by lpil** [comment]
I'm afraid as this is a core feature I need to maintain I'm going to be taking the lead on this one so that I have full understanding of how it works. Not knowing how part of the compiler works puts me in an tricky position for future work and making informed decisions about it's development.

**Comment by gamebox** [comment]
Oh, ok. 👍

**Comment by lpil** [comment]
Having said that it would be good to see any work you have done. I'm sure we can use it so long as I'm able to get a good understanding of how LSP and our implementation works.

**Comment by lpil** [comment]
https://rust-analyzer.github.io/blog/2020/07/20/three-architectures-for-responsive-ide.html

**Comment by adituv** [comment]
I'm looking at making a start on implementing this.  My current plan is to use the [lsp-server](https://github.com/rust-analyzer/lsp-server) crate as a base, then build on top of that.

The rust-analyzer tool, which builds on top of lsp-server, mentions a [major performance degradation](https://github.com/rust-analyzer/rust-analyzer/blob/28b1cd32b3682f172d9caa862dbf1bcdbf53234b/crates/rust-analyzer/src/main_loop.rs#L31) on Windows, so we may want to implement a similar workaround.

**Comment by lpil** [comment]
That's a very interesting looking crate! Thanks for the recommendation.

I'd like to suggest implementing formatting first in the LSP as that's probably the simplest thing we can do in the compiler at present. It's important that it's clear how to work on and maintain this feature (for me especially) so it would be good to do this in small incremental pieces if possible.

**Comment by natdm** [comment]
This is something I'd love to help with when the first iteration is done. What's the status?

**Comment by lpil** [comment]
Hello! No work has been done at present. I am interested in implementing this using just an event loop, IO, and Rust's serde JSON rather than a generic library (at least to start).

I think that implementing formatting via LSP could be a good place to start.

**Comment by natdm** [comment]
Another thing to consider is treesitter support, further on down the road. Will give you tons of functionality and syntax highlighting just by having the grammar and queries (neovim only).

**Comment by oblitum** [comment]
On LSP sense, implement semantic tokens can have more impact than tree-sitter, for which a parser could be provided by 3rd party efforts. On Vim/NeoVim's side coc.nvim can already leverage tree-sitter colorschemes and make them work for servers that provide LSP semantic tokens. It would also benefit any other editor, VSCode, etc.

**Comment by HarryET** [comment]
Currently started an implementation for the VSCode extension - can be move to its own repo when working? - it has an open draft PR here: https://github.com/gleam-lang/vscode-gleam/pull/19

**Comment by lpil** [comment]
Closing this in favour of the LSP milestone

https://github.com/gleam-lang/gleam/milestone/20
